### PR TITLE
Hi mentors!

### PR DESCRIPTION
**Pull request by nanemo**
Web application is created. Now i am working on Session, Servlet Testing and CSS. Application works on 14 JDK and on Tomcat 9.0.48 version. Program works through JDBC. You can see all abst classes about DAO in service package.

I will check all my code for neatness at the end via SonarLint )).

### CODE REVIEW

**Comment by nanemo** [comment]
Yuriy and Andrey, my web app works via JDBC, but i have a problem. I don't know how to push MySql tables with project code to GitHub that you could run project together both of them.

**Comment by nanemo** [comment]
image

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <junit.version>5.8.2</junit.version>
+    </properties>
+
+    <dependencies>
+
+        <dependency>
+            <groupId>javax.servlet</groupId>
+            <artifactId>javax.servlet-api</artifactId>
+            <version>4.0.1</version>
```
Comment (Anderwerty) [inline pom.xml]: please extract this version into maven properties
the same about other dependencies

File: pom.xml [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+            <artifactId>lombok</artifactId>
+            <version>1.18.24</version>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>mysql</groupId>
+            <artifactId>mysql-connector-java</artifactId>
+            <version>8.0.28</version>
+        </dependency>
+        <dependency>
```
Comment (Anderwerty) [inline pom.xml]: do we need it as dependency?

File: src/main/java/ru/javarush/quest/controller/LogicServlet.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    private static int count = 0;
+    private final Context context = Context.getDaoControllerInstance();
+
+    private static synchronized void getCount() {
+        count++;
+    }
+
+    @Override
+    protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+        if (count > 3) count = 0;
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/controller/LogicServlet.java]: `3` is magic number

File: src/main/java/ru/javarush/quest/controller/LogicServlet.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            req.getRequestDispatcher("errorPage.jsp").forward(req, resp);
+            count = 0;
+        }
+    }
+
+    private Answer getAnswer() {
+        return context.getAnswer(count);
+    }
+
+    private Question getQuestion() {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/controller/LogicServlet.java]: logic should be in service

File: src/main/java/ru/javarush/quest/entity/Account.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
+package ru.javarush.quest.entity;
+
+public class Account {
+    private Long id;
+    private final String name;
+    private final String login;
+    private String password;
+
+    public Account(Long id, String name, String login, String password) {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/entity/Account.java:12]: We have 4 fields. Let's make all fields final and use builder pattern

File: src/main/java/ru/javarush/quest/entity/Account.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        return login;
+    }
+
+    public void setId(Long id) {
+        this.id = id;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/entity/Account.java]: https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used
Braces are used with if, else, for, do and while statements, even when the body is empty or contains only a single statement.

File: src/main/java/ru/javarush/quest/entity/Account.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        Account account = (Account) o;
+
+        if (!id.equals(account.id)) return false;
+        if (!name.equals(account.name)) return false;
+        if (!login.equals(account.login)) return false;
+        return password.equals(account.password);
+    }
+
+    @Override
+    public int hashCode() {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/entity/Account.java]: let's use `Objects.hash` for simplification

File: src/main/java/ru/javarush/quest/entity/Account.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
+package ru.javarush.quest.entity;
+
+public class Account {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/entity/Account.java:3]: You have already added lombok. Why not use it for toString, equals and hashCode methods generation

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+import ru.javarush.quest.entity.Answer;
+import ru.javarush.quest.service.AbstractDao;
+import ru.javarush.quest.service.AnswerDao;
+
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+
+public class AnswerImpl extends AbstractDao implements AnswerDao {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: Maybe it is dao?

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public Answer getAnswersInOrder(int count) {
+
+        Answer answer = null;
+        try (Connection connection = connect();
+             Statement stmt = connection.createStatement()) {
+            stmt.execute("SELECT * from answers where id=" + count);
+            ResultSet resultSet = stmt.getResultSet();
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: it seems that https://docs.oracle.com/javase/7/docs/api/java/sql/ResultSet.html implements Autoclosable

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        try (Connection connection = connect();
+             Statement stmt = connection.createStatement()) {
+            stmt.execute("SELECT d.description FROM description_for_wrong_answer d LEFT JOIN answers " +
+                    "ON answers.id = d.id where d.id = " + count);
+            ResultSet rs = stmt.getResultSet();
+            while (rs.next()) {
+                wrongAnswerDescription = getWrongAnswerDescription(rs);
+            }
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: It is ok just log exception and return null?

File: src/main/java/ru/javarush/quest/impl/LoginImpl.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
+package ru.javarush.quest.impl;
+
+import ru.javarush.quest.entity.Account;
+import ru.javarush.quest.service.AbstractDao;
+import ru.javarush.quest.service.LoginDao;
+
+import java.sql.*;
+
+
+public class LoginImpl extends AbstractDao implements LoginDao {
```
Comment (Anderwerty) [inline src/main/java/ru/javarush/quest/impl/LoginImpl.java]: the same

**Review by Anderwerty — COMMENTED** [review]
It is also expected to write tests for your dao layer

File: docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,30 @@
+version:
+services:
```
Comment (Anderwerty) [inline docker-compose.yml:3]: Sorry, but what functionality will create a database structure?

File: pom.xml (hunk 1 above)
Comment (nanemo) [reply pom.xml]: problem solved

File: pom.xml (hunk 2 above)
Comment (nanemo) [reply pom.xml]: deleted

File: src/main/java/ru/javarush/quest/controller/LogicServlet.java (hunk 3 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/controller/LogicServlet.java]: problem solved. for number 3 i added context.getQuestionSize() method. It is game level size.

File: src/main/java/ru/javarush/quest/controller/LogicServlet.java (hunk 4 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/controller/LogicServlet.java]: problem solved

Comment (nanemo) [reply src/main/java/ru/javarush/quest/controller/LogicServlet.java]: I took both of them to Contex class

File: src/main/java/ru/javarush/quest/entity/Account.java (hunk 6 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/entity/Account.java]: i deleted this equals method.

File: src/main/java/ru/javarush/quest/entity/Account.java (hunk 7 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/entity/Account.java]: i deleted this hash method.

File: src/main/java/ru/javarush/quest/entity/Account.java (hunk 8 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/entity/Account.java:3]: I deleted lombok. I dont use already.

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java (hunk 9 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: I replaced all names about Impl classes to Dao.

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java (hunk 10 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: ResultSet close method was added in finally block

File: src/main/java/ru/javarush/quest/impl/AnswerImpl.java (hunk 11 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/impl/AnswerImpl.java]: What do you here. i didnt get it

File: src/main/java/ru/javarush/quest/impl/LoginImpl.java (hunk 12 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/impl/LoginImpl.java]: replaced

File: docker-compose.yml (hunk 13 above)
Comment (nanemo) [reply docker-compose.yml:3]: Please look at here again. About docker helped to me from our group Marina.

File: src/main/java/ru/javarush/quest/entity/Account.java (hunk 5 above)
Comment (nanemo) [reply src/main/java/ru/javarush/quest/entity/Account.java:12]: What is the different will be if i add builder pattern here? and why i have to add just in Account class no in Answer or Question classes?

**Comment by nanemo** [comment]
You can check my last commit in dev branch.

**Pull request merged**